### PR TITLE
feat(examples): remote exex

### PR DESCRIPTION
**Pull request by shekhirin**
- `cargo run --bin exex -- node --chain holesky --debug.etherscan` on the left
- `cargo run --bin consumer` on the right

image

---

The protobuf conversion implementations are kinda insane just for the sake of example, but I think they may be handy for others to use. We may move them into some crate in the future.

### CODE REVIEW

**Review by emhane — COMMENTED** [review]
would be nice if can align more with rust's standard interfaces for type conversions `From`, `Into`, `FromIterator`, `IntoIterator`

File: examples/exex/remote/src/codec.rs [hunk 1]
```diff
@@ -0,0 +1,822 @@
...
+        timestamp: header.timestamp,
+        mix_hash: header.mix_hash.to_vec(),
+        nonce: header.nonce,
+        base_fee_per_gas: header.base_fee_per_gas,
+        blob_gas_used: header.blob_gas_used,
+        excess_blob_gas: header.excess_blob_gas,
+        parent_beacon_block_root: header.parent_beacon_block_root.map(|root| root.to_vec()),
+        extra_data: header.extra_data.to_vec(),
+    }
+}
```
[the remark this answers is not in the document]
Comment (shekhirin) [reply examples/exex/remote/src/codec.rs]: yep, was faster to do it via free standing functions but agreed we should do it properly via two `From` traits for `Reth -> Protobuf` and `Protobuf -> Reth` conversions

File: examples/exex/remote/src/codec.rs [hunk 2]
```diff
@@ -0,0 +1,822 @@
...
+use eyre::OptionExt;
+use reth::primitives::{Address, BlockHash, Bloom, TxHash, B256, U256};
+
+use crate::proto;
+
+pub fn to_proto_notification(
+    notification: &reth_exex::ExExNotification,
+) -> eyre::Result<proto::ExExNotification> {
+    let notification = match notification {
+        reth_exex::ExExNotification::ChainCommitted { new } => {
```
[the remark this answers is not in the document]
Comment (shekhirin) [reply examples/exex/remote/src/codec.rs]: yep, many names are the same between Reth types and protobuf-generated types

File: examples/exex/remote/src/codec.rs [hunk 3]
```diff
@@ -0,0 +1,822 @@
...
+        blocks: chain
+            .blocks_iter()
+            .map(|block| proto::Block {
+                header: Some(proto::SealedHeader {
+                    hash: block.header.hash().to_vec(),
+                    header: Some(to_proto_header(&block.header.header())),
+                }),
+                body: block.transactions().map(to_proto_transaction).collect(),
+                ommers: block.ommers.iter().map(to_proto_header).collect(),
+                senders: block.senders.iter().map(|sender| sender.to_vec()).collect(),
```
[the remark this answers is not in the document]
Comment (shekhirin) [reply examples/exex/remote/src/codec.rs]: yep, done!

File: examples/exex/remote/src/codec.rs [hunk 4]
```diff
@@ -0,0 +1,822 @@
...
+                                    key: key.to_le_bytes_vec(),
+                                    previous_or_original_value: slot
+                                        .previous_or_original_value
+                                        .to_le_bytes_vec(),
+                                    present_value: slot.present_value.to_le_bytes_vec(),
+                                })
+                                .collect(),
+                            status: to_proto_account_status(account.status) as i32,
+                        })
+                    })
```
[the remark this answers is not in the document]
Comment (shekhirin) [reply examples/exex/remote/src/codec.rs]: it's not just `From`, it's `TryFrom`. so instead I did `TryFrom<(Address, proto::Account)> for reth::BundleAccount` and just manually mapped

File: examples/exex/remote/bin/exex.rs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+use reth_exex::{ExExContext, ExExEvent, ExExNotification};
+use reth_node_api::FullNodeComponents;
+use reth_node_ethereum::EthereumNode;
+use tokio::sync::{broadcast, mpsc};
+use tokio_stream::wrappers::ReceiverStream;
+use tonic::{transport::Server, Request, Response, Status};
+
+#[derive(Debug)]
+struct ExExService {
+    notifications: Arc<broadcast::Sender<ExExNotification>>,
```
Comment (mattsse) [inline examples/exex/remote/bin/exex.rs:16]: broadcast sender is already clone and just an arc wrapper

**Review by mattsse — APPROVED** [review]
this adds a lot of deps but should be okay for the example

I assume we could perhaps provide the proto bindings separately

**Pull request merged**